lib/reek/smells/attribute.rb

```diff
@@ -42,14 +52,37 @@ def examine_context(ctx)
       private
 
       def attributes_in(module_ctx)
-        result = Set.new
-        attr_defn_methods = [:attr, :attr_reader, :attr_writer, :attr_accessor]
         module_ctx.local_nodes(:send) do |call_node|
-          if attr_defn_methods.include?(call_node.method_name)
-            call_node.arg_names.each { |arg| result << [arg, call_node.line] }
+          if visibility_modifier?(call_node)
+            track_visibility(call_node)
+          elsif ATTR_DEFN_METHODS.include?(call_node.method_name)
+            call_node.arg_names.each do |arg|
+              @visiblity_tracker[arg] = @visiblity_mode
+              @result << [arg, call_node.line]
+            end
           end
         end
-        result
+        @result.select { |args| public_methods_from_tracker.include?(args[0]) }
+      end
+
+      def visibility_modifier?(call_node)
+        VISIBILITY_MODIFIERS.include?(call_node.method_name)
+      end
+
+      def track_visibility(call_node)
+        if modifier_with_attributes?(call_node)
+          call_node.arg_names.each { |arg| @visiblity_tracker[arg] = call_node.method_name }
+        else
+          @visiblity_mode = call_node.method_name
+        end
+      end
+
+      def public_methods_from_tracker
```

Comment:
You abstract the "visibility mode" you're interested away by using VISIBILITY_MODE. But in this method name you kind of make this abstraction useless since it explicitly says "public". Maybe this abstraction is over the top? If you just care about not "fat fingering" "public" and still have an abstraction, how about renaming VISIBILITY_MODE TO PUBLIC_VISIBILITY?
Also I'd prefer to drop the "from_tracker" part in the method name since this is an unncessary detail.
So I'd just call this "public_methods".
WDYT?